Lib/asyncio/timeouts.py

```diff
@@ -0,0 +1,158 @@
+import contextlib
+import enum
+
+from types import TracebackType
+from typing import final, Any, Iterator, Optional, Type
+
+from . import events
+from . import tasks
+
+
+__all__ = (
+    "cancel_scope",
+    "cancel_scope_at",
+    "timeout",
+    "timeout_at",
+    "CancelScope"
+)
+
+
+class _State(enum.Enum):
+    CREATED = "created"
+    ENTERED = "active"
+    CANCELLED = "cancelled"
+    EXITED = "exited"
+
+
+@final
+class CancelScope:
+    __slots__ = ("_deadline", "_loop", "_state", "_timeout_handler")
+
+    def __init__(
+        self, deadline: Optional[float], loop: events.AbstractEventLoop
+    ) -> None:
+        self._loop = loop
+        self._state = _State.CREATED
+
+        self._timeout_handler: Optional[events.TimerHandle] = None
+        self._deadline = deadline
+        self._reschedule()
+
+    @property
+    def deadline(self) -> Optional[float]:
+        return self._deadline
+
+    @deadline.setter
+    def deadline(self, value: Optional[float]) -> None:
+        self._deadline = value
+        self._reschedule()
+
+    def cancelled(self) -> bool:
```

Comment:
trio-style has 'cancel_called'. The name is not in line with asyncio style.

asyncio uses `cancelled()` for tasks and futures, let's reuse it.